Crust/Mapper/Mapper.swift

```diff
@@ -32,48 +32,76 @@ public struct Mapper<T: Mapping> {
+        
+        // TODO: Figure out better ways to represent `nil` keyPaths than `""`.
+        let baseJson = json[spec.keyPath] ?? json
+        
+        var object = try { () -> T.MappedObject in
+            let mapping = spec.mapping
+            guard let object = try mapping.getExistingInstance(json: baseJson) else {
+                return try mapping.getNewInstance()
+            }
+            return object
+        }()
+        
+        let context = MappingContext(withObject: object, json: baseJson, direction: MappingDirection.fromJSON)
+        context.parent = parentContext
+        try self.perform(spec, on: &object, with: context)
+        
+        return object
+    }
+    
     public func map(from json: JSONValue, using mapping: T, parentContext: MappingContext? = nil) throws -> T.MappedObject {
         let object = try { () -> T.MappedObject in
             guard let object = try mapping.getExistingInstance(json: json) else {
                 return try mapping.getNewInstance()
             }
             return object
         }()
-        return try mapFromJSON(json, toObject: object, mapping: mapping, parentContext: parentContext)
+        return try map(from: json, to: object, using: mapping, parentContext: parentContext)
     }
     
-    public func mapFromJSON(_ json: JSONValue, toObject object: T.MappedObject, mapping: T, parentContext: MappingContext? = nil) throws -> T.MappedObject {
+    public func map(from json: JSONValue, to object: T.MappedObject, using mapping: T, parentContext: MappingContext? = nil) throws -> T.MappedObject {
         var object = object
         let context = MappingContext(withObject: object, json: json, direction: MappingDirection.fromJSON)
         context.parent = parentContext
-        try mapping.performMappingWithObject(&object, context: context)
+        try self.perform(mapping, on: &object, with: context)
         return object
     }
     
     public func mapFromObjectToJSON(_ object: T.MappedObject, mapping: T) throws -> JSONValue {
         var object = object
         let context = MappingContext(withObject: object, json: JSONValue.object([:]), direction: MappingDirection.toJSON)
-        try mapping.performMappingWithObject(&object, context: context)
+        try self.perform(mapping, on: &object, with: context)
         return context.json
     }
+    
+    internal func perform(_ mapping: T, on object: inout T.MappedObject, with context: MappingContext) throws {
+        try mapping.start(context: context)
+        mapping.execute(object: &object, context: context)
+        try mapping.complete(object: &object, context: context)
+    }
+    
+    internal func perform(_ spec: Spec<T>, on object: inout T.MappedObject, with context: MappingContext) throws {
+        let mapping = spec.mapping
+        try mapping.start(context: context)
+        object <- (spec, context)
```

Comment:
this says, get the mapping. start mapping (which starts a transaction with the DB). Do the mapping.